libraries/chain/fork_database.cpp

```diff
@@ -57,48 +70,135 @@ namespace eosio { namespace chain {
 
       auto fork_db_dat = my->datadir / config::forkdb_filename;
       if( fc::exists( fork_db_dat ) ) {
-         string content;
-         fc::read_file_contents( fork_db_dat, content );
-
-         fc::datastream<const char*> ds( content.data(), content.size() );
-         unsigned_int size; fc::raw::unpack( ds, size );
-         for( uint32_t i = 0, n = size.value; i < n; ++i ) {
-            block_state s;
-            fc::raw::unpack( ds, s );
-            set( std::make_shared<block_state>( move( s ) ) );
-         }
-         block_id_type head_id;
-         fc::raw::unpack( ds, head_id );
+         try {
+            string content;
+            fc::read_file_contents( fork_db_dat, content );
+
+            fc::datastream<const char*> ds( content.data(), content.size() );
+
+            // validate totem
+            uint32_t totem = 0;
+            fc::raw::unpack( ds, totem );
+            EOS_ASSERT( totem == magic_number, fork_database_exception,
+                        "Fork database file '${filename}' has unexpected magic number!",
+                        ("filename", fork_db_dat.generic_string())
+            );
+
+            // validate version
+            uint32_t version = 0;
+            fc::raw::unpack( ds, version );
+            EOS_ASSERT( version >= min_supported_version && version <= max_supported_version,
+                        fork_database_exception,
+                       "Unsupported version of fork database file '${filename}'. "
+                       "Fork database version is ${version} while code supports version(s) [${min},${max}]",
+                       ("filename", fork_db_dat.generic_string())
+                       ("version", version)
+                       ("min", min_supported_version)
+                       ("max", max_supported_version)
+            );
+
+            block_header_state bhs;
+            fc::raw::unpack( ds, bhs );
+            reset( bhs );
+
+            unsigned_int size; fc::raw::unpack( ds, size );
+            for( uint32_t i = 0, n = size.value; i < n; ++i ) {
+               block_state s;
+               fc::raw::unpack( ds, s );
+               for( const auto& receipt : s.block->transactions ) {
+                  if( receipt.trx.contains<packed_transaction>() ) {
+                     auto& pt = receipt.trx.get<packed_transaction>();
+                     s.trxs.push_back( std::make_shared<transaction_metadata>( std::make_shared<packed_transaction>(pt) ) );
```

Comment:
Would be nice to have a test that failed before that now works for transactions in fork_db. Maybe this is something @brianjohnson5972 or @taokayan could create.